web-components/src/components/form/BasketTakeForm.tsx

```diff
@@ -0,0 +1,224 @@
+ *
+ * Also:
+ * https://buf.build/regen/regen-ledger/docs/main:regen.ecocredit.basket.v1#regen.ecocredit.basket.v1.EventTake
+ */
+
+const useStyles = makeStyles((theme: Theme) => ({
+  checkboxLabel: {
+    marginTop: theme.spacing(10.75),
+    alignItems: 'initial',
+    '& .MuiCheckbox-root': {
+      alignSelf: 'end',
+    },
+  },
+  checkboxDescription: {
+    color: theme.palette.primary.contrastText,
+    fontSize: theme.spacing(4.5),
+    fontWeight: 'bold',
+    marginLeft: theme.spacing(1),
+  },
+}));
+
+// Output (submit)
+export interface MsgTakeValues {
+  owner: string;
+  basketDenom: string;
+  amount: string;
+  retireOnTake: boolean;
+  retirementLocation?: string;
+  retirementNote?: string;
+}
+
+interface CreditTakeFormValues {
+  amount: number;
+  retireOnTake?: boolean;
+  note?: string;
+  country: string;
+  stateProvince: string;
+  postalCode?: string;
+  retirementLocation?: string;
+}
+
+// Input (args)
+interface FormProps {
+  mapboxToken: string;
+  accountAddress: string;
+  basket: Basket;
+  availableTradableAmount: number;
+  onClose: () => void;
+  onSubmit: (values: MsgTakeValues) => void;
+}
+
+const BasketTakeForm: React.FC<FormProps> = ({
+  mapboxToken,
+  accountAddress,
+  basket,
+  availableTradableAmount,
+  onClose,
+  onSubmit,
+}) => {
+  const styles = useStyles();
+
+  const initialValues = {
+    amount: 0,
+    retireOnTake: !basket.disableAutoRetire,
+    note: '',
+    country: 'US',
+    stateProvince: '',
+    retirementLocation: undefined,
+  };
+
+  const validateHandler = (
+    values: CreditTakeFormValues,
+  ): FormikErrors<CreditTakeFormValues> => {
+    let errors: FormikErrors<CreditTakeFormValues> = {};
+
+    if (!values.amount) {
+      errors.amount = requiredMessage;
+    } else if (Math.sign(values.amount) !== 1) {
+      errors.amount = invalidAmount;
+    } else if (values.amount > availableTradableAmount) {
+      errors.amount = insufficientCredits;
+    }
+
+    // Retire form validation (optional subform)
+    if (values.retireOnTake) {
+      const retirementValues: RetireFormValues = {
+        retiredAmount: values.amount,
+        note: values.note || '',
+        country: values.country,
+        stateProvince: values.stateProvince,
+        postalCode: values.postalCode,
+      };
+      errors = validateCreditRetire(
+        availableTradableAmount,
+        retirementValues,
+        errors,
+      );
+    }
+    return errors;
+  };
+
+  const submitHandler = async (values: CreditTakeFormValues): Promise<void> => {
+    const msgTake: MsgTakeValues = {
+      owner: accountAddress,
+      basketDenom: basket.basketDenom,
+      amount: (values.amount * Math.pow(10, basket.exponent)).toString(),
+      retireOnTake: !!values.retireOnTake,
+      retirementLocation: values.retirementLocation,
+      retirementNote: values?.note,
+    };
+
+    onSubmit(msgTake);
+  };
+
+  const AutoSetRetirementLocation = (): JSX.Element => {
```

Comment:
curious why this is a component instead of a normal effect. Is it so you have access to formik context?

Reply:
Yes, I pulled this from the Formik docs. I believe the hook has to be inside the <Form> for the context to work.